WalletWasabi/Wallets/TransactionFeeProvider.cs

```diff
@@ -0,0 +1,136 @@
+using System.Threading.Channels;
+using System.Threading.Tasks;
+using WalletWasabi.Blockchain.Transactions;
+using WalletWasabi.Logging;
+using WalletWasabi.Tor.Http;
+using WalletWasabi.Tor.Http.Extensions;
+using WalletWasabi.WebClients.Wasabi;
+
+namespace WalletWasabi.Wallets;
+
+public class TransactionFeeProvider : BackgroundService
+{
+	private const int MaximumDelayInSeconds = 120;
+	private const int MaximumRequestsInParallel = 3;
+
+	public TransactionFeeProvider(WasabiHttpClientFactory httpClientFactory)
+	{
+		HttpClient = httpClientFactory.NewHttpClient(httpClientFactory.BackendUriGetter, Tor.Socks5.Pool.Circuits.Mode.NewCircuitPerRequest);
+	}
+
+	public event EventHandler<EventArgs>? RequestedFeeArrived;
+
+	public ConcurrentDictionary<uint256, Money> FeeCache { get; } = new();
+	public ConcurrentQueue<uint256> Queue { get; } = new();
+	private SemaphoreSlim Semaphore { get; } = new(initialCount: 0, maxCount: MaximumRequestsInParallel);
+	private IHttpClient HttpClient { get; }
+
+	private async Task FetchTransactionFeeAsync(uint256 txid, CancellationToken cancellationToken)
+	{
+		const int MaxAttempts = 3;
+
+		for (int i = 0; i < MaxAttempts; i++)
+		{
+			try
+			{
+				using CancellationTokenSource timeoutCts = new(TimeSpan.FromSeconds(20));
+				using CancellationTokenSource linkedCts = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken, timeoutCts.Token);
+
+				var response = await HttpClient.SendAsync(
+					HttpMethod.Get,
+					$"api/v{Helpers.Constants.BackendMajorVersion}/btc/Blockchain/get-transaction-fee?transactionId={txid}",
+					null,
+					linkedCts.Token).ConfigureAwait(false);
+
+				if (response.StatusCode != HttpStatusCode.OK)
+				{
+					await response.ThrowRequestExceptionFromContentAsync(cancellationToken).ConfigureAwait(false);
+				}
+
+				Money fee = await response.Content.ReadAsJsonAsync<Money>().ConfigureAwait(false);
+
+				if (!FeeCache.TryAdd(txid, fee))
+				{
+					throw new InvalidOperationException($"Failed to cache {txid} with fee: {fee}");
+				}
+
+				RequestedFeeArrived?.Invoke(this, EventArgs.Empty);
+				return;
+			}
+			catch (Exception ex)
+			{
+				if (cancellationToken.IsCancellationRequested)
+				{
+					return;
+				}
+
+				Logger.LogWarning($"Attempt: {i}. Failed to fetch transaction fee. {ex}");
+			}
+		}
+	}
+
+	public bool TryGetFeeFromCache(uint256 txid, [NotNullWhen(true)] out Money? fee)
+	{
+		return FeeCache.TryGetValue(txid, out fee);
+	}
+
+	public void BeginRequestTransactionFee(SmartTransaction tx)
+	{
+		if (!tx.Confirmed && tx.ForeignInputs.Count != 0)
+		{
+			Queue.Enqueue(tx.GetHash());
+			Semaphore.Release(1);
+		}
+	}
+
+	protected override async Task ExecuteAsync(CancellationToken cancel)
+	{
+		while (!cancel.IsCancellationRequested)
+		{
+			await Semaphore.WaitAsync(cancel).ConfigureAwait(false);
+
+			if (!Queue.TryDequeue(out var txidToFetch))
+			{
+				continue;
+			}
+
+			try
+			{
+				_ = ScheduledTask(txidToFetch);
```

Comment:
Whenever a Task throws an exception and it is not read from task results before the ThreadPool or GC collects it, you will get an unobserverException. It is not fatal but should be avoided. Here you do not await the task so this line will never throw. You have to move the exception handling inside.

Reply:
True, brain fart.